File: src/itemview.ts
"use strict";

import * as _ from "underscore";
import * as Handlebars from "handlebars";

export class AItemView {
	public wasm_module: any = null;

	/**
	 * Конструктор представления.
	 * 
	 * @method constructor
	 * @param {any} wasm_context Контекст.
	 * @param {any} wasm_parent Родительский узел.
	 * @param {string} nodeId Уникальный идентификатор узла.
	 * @param {any} nodeCreateInfo Свойства представления.
	 */
	constructor(public wasm_context: any = null, public wasm_parent: any = null, nodeId: string, nodeCreateInfo?: any) {
		this.wasm_module = function (self: any) {
			var AItemViewObject = wasm_context.module.AItemView.extend("AItemView", {
				__construct: function(parent: any, nodeId: string, nodeCreateInfo: any) {
					this.__parent.__construct.call(this, parent, nodeId, _.defaults(nodeCreateInfo, {
						stylesheet: null,
						classes: []
					}));
				},
				__destruct: function() {
					this.__parent.__destruct.call(this);
				},
				initialize: _.bind(self.initialize, self),
				update: _.bind(self.update, self)
			});

			return new AItemViewObject(wasm_parent, nodeId, nodeCreateInfo);
		}(this) as any;
	}

	public initialize(): void {
		// Empty
	}

	public update(): void {
		// Empty
	}
};

File: lib/src/itemview.cpp
#include <sway/webcore/nodeelementupdater.hpp>

#include <sway/webcore/mvc/itemview.h>

namespace sway::webcore::mvc {

EMSCRIPTEN_BINDING_BEGIN(AItemView) {
  emscripten::class_<AItemView, emscripten::base<AView>>("AItemView")
      .allow_subclass<AItemViewWrapper>(
          "AItemViewWrapper", emscripten::constructor<core::NodePtr_t, std::string, TreeNodeElementDescriptor>())
      .constructor<core::NodePtr_t, std::string, TreeNodeElementDescriptor>()
      .function(
          "initialize", emscripten::optional_override([](AItemView &self) { return self.AItemView::initialize(); }))
      .function("update", emscripten::optional_override([](AItemView &self) { return self.AItemView::update(); }));
}
EMSCRIPTEN_BINDING_END()

AItemView::AItemView(core::NodePtr_t parent,
    // const core::NodeIndex & nodeIndex,
    const std::string &nodeId, const TreeNodeElementDescriptor &createInfo)
    : AView(parent, core::NodeIndex(), nodeId, createInfo) {
  // Empty
}

AItemView::AItemView(core::NodePtr_t parent, const core::NodeIndex &nodeIndex, const std::string &nodeId,
    const TreeNodeElementDescriptor &createInfo)
    : AView(parent, nodeIndex, nodeId, createInfo) {
  // Empty
}

auto AItemView::traverse(core::typedefs::TraverserSharedPtr_t traverser) -> u32_t {
  return core::toBase(core::TraverserAction::Enum::NONE);
}
// void AItemView::accept(ITreeVisitor * visitor) {
// 	visitor->visitOnEnter(this);

// 	for (core::NodePtr_t node : getChildren())
// 		static_cast<AItemView *>(node)->accept(visitor);
// }

void AItemView::initialize() {
  // Empty
}

void AItemView::update() {
  // EM_ASM({console.log("ON_ENTRY_CHANGED " + UTF8ToString($0))}, getNodeId().c_str());
}

void AItemViewWrapper::initialize() { return call<void>("initialize"); }

void AItemViewWrapper::update() { return call<void>("update"); }

}  // namespace sway::webcore::mvc
